sstables: introduce trie_writer

michoecho opened the pull request:
This is the first part of a larger project meant to implement a trie-based
index format. (The same or almost the same as Cassandra's BTI).

As of this patch, the new code isn't used for anything yet,
but we introduced separately from its users to keep PRs small enough
for reviewability.

This commit introduces trie_writer, a class responsible for turning a
stream of (key, value) pairs (already sorted by key) into a stream of
serializable nodes, such that:

1. Each node lies entirely within one page (guaranteed).
2. Parents are located in the same page as their children (best-effort).
3. Padding (unused space) is minimized (best-effort).

It does mostly what you would expect a "sorted keys -> trie" builder to do.
The hard part is calculating the sizes of nodes (which, in a well-packed on-disk
format, depend on the exact offsets of the node from its children) and grouping
them into pages.

This implementation mostly follows Cassandra's design of the same thing.
There are some differences, though. Notable ones:

1. The writer operates on chains of characters, rather than single characters.

   In Cassandra's implementation, the writer creates one node per character.
   A single long key can be translated to thousands of nodes.
   We create only one node per key. (Actually we split very long keys into
   a few nodes, but that's arbitrary and beside the point).

   For BTI's partition key index this doesn't matter.
   Since it only stores a minimal unique prefix of each key,
   and the trie is very balanced (due to token randomness),
   the average number of new characters added per key is very close to 1 anyway.
   (And the string-based logic might actually be a small pessimization, since
   manipulating a 1-byte string might be costlier than manipulating a single byte).

   But the row index might store arbitrarily long entries, and in that case the
   character-based logic might result in catastrophically bad performance.
   For reference: when writing a partition index, the total processing cost
   of a single node in the trie_writer is on the order of 800 instructions.
   Total processing cost of a single tiny partition during a `upgradesstables`
   operation is on the order of 10000 instructions. A small INSERT is on the
   order of 40000 instructions.

   So processing a single 1000-character clustering key in the trie_writer
   could cost as much as 20 INSERTs, which is scary. Even 100-character keys
   can be very expensive. With extremely long keys like that, the string-based
   logic is more than ~100x cheaper than character-based logic.
   (Note that only *new* characters matter here. If two index entries share a
   prefix, that prefix is only processed once. And the index is only populated
   with the minimal prefix needed to distinguish neighbours. So in practice,
   long chains might not happen often. But still, they are possible).

   I don't know if it makes sense to care about this case, but I figured the
   potential for problems is too big to ignore, so I switched to chain-based logic.

2. In the (assumed to be rare) case when a grouped subtree turns out to be bigger
   than a full page after revising the estimate, Cassandra splits it in a
   different way than us.

For testability, there is some separation between the logic responsible
for turning a stream of keys into a stream of nodes, and the logic
responsible for turning a stream of nodes into a stream of bytes.
This commit only includes the first part. It doesn't implement the target
on-disk format yet.

The serialization logic is passed to trie_writer via a template parameter.

There is only one test added in this commit, which attempts to be exhaustive,
by testing all possible datasets up to some size. The run time of the test
grows exponentially with the parameter size. I picked a set of parameters
which runs fast enough while still being expressive enough to cover all
the logic. (I checked the code coverage). But I also tested it with greater parameters
on my own machine (and with DEVELOPER_BUILD enabled, which adds extra sanitization).

Refs scylladb/scylladb#19191

New functionality, no backporting needed.

michoecho commented:
@denesb @avikivity Reviewers needed. Will you be willing to review this? If not, who could?

avikivity commented:
> @denesb @avikivity Reviewers needed. Will you be willing to review this? If not, who could?

Reviewing.

avikivity commented on configure.py:
```diff
@@ -863,6 +864,7 @@ def find_ninja():
                 'sstables/random_access_reader.cc',
                 'sstables/metadata_collector.cc',
                 'sstables/writer.cc',
+                'sstables/trie/trie_writer.cc',
```
Need cmake equivalent.

michoecho replied:
Fixed.

avikivity commented on sstables/trie/trie_writer.hh:
Can we integrate this with asan? There are hooks to poison/unpoison memory, and we can leave poisoned gaps in debug mode. Then pointer debugging will be done by asan.

michoecho commented on sstables/trie/trie_writer.cc:
```diff
@@ -0,0 +1,11 @@
+
+#include <seastar/util/log.hh>
+
+seastar::logger trie_logger("trie");
```
Ideally, I guess there should be a header with declarations only, a header with templated implementations, a header with explicit instantiation declarations, a translation unit with explicit instantiations and non-templated implementations. I didn't bother though.

avikivity replied:
Okay, big template for testing.

avikivity replied:
> Ideally, I guess there should be a header with declarations only, a header with templated implementations, a header with explicit instantiation declarations, a translation unit with explicit instantiations and non-templated implementations. I didn't bother though.

Not so important for a header which will be included from few translation units.

avikivity commented on sstables/trie/trie_writer.hh:
Likely faster to do a size-independent compare (if trailer is consistently initialized).

michoecho replied:
The only reason to compare payloads is testing. This won't have any use outside of tests.

avikivity commented on sstables/trie/trie_writer.hh:
Shouldn't it be explicit? implicit conversion undoes strong typing.

michoecho replied:
Not in this case, this was deliberate. `node_size` is in some sense a subtype of `sink_offset`.

avikivity commented on sstables/trie/trie_writer.hh:
Would it make sense to allocate it locally here (and limit it to some small amount?)

avikivity commented on sstables/trie/trie_writer.hh:
Will _children_capacity quickly grow to 256?

Why not always make it 256 (and index)? These are reclaimed quickly, no?

I see _children as having its leftmost populated branch deeply nested (with the leftmost child's leftmost branch being the most deeply populated), with the rightmost children being not deeply nested. Is this a correct picture?

michoecho replied:
If you made it always 256, you would grow all the small nodes to at least 2 kiB in size. I'm not sure that's a better idea than resizing the vector.

michoecho replied:
> Is it correct that we wait to flush until we think we can fill a page? Is this the only condition where we avoid flusing a completed subtree?

It's not that we "avoid" flushing a completed subtree. The whole point of all logic added in this PR is trying to keep children in the same page as their parents. And that is done by accumulating subtrees until they can't fit in a page anymore, and only splitting them when that happens.

avikivity replied:
> The reason not to flush A1 is that A1 might fit on one page with A. If you flush A1 now, and A2, A3, A4, ... have subtrees larger than a page, then you guarantee that A1 will be on a different page than A.
> 

But, A2 and A3 will not be co-located with A.

Statically, A2 and A3 have a higher probability of being used than A1 (since they are larger). Dynamically of course it might be different.


> If you flush nodes immediately after they are completed, they are written to the file in postorder. I.e. <img alt="image" width="300" height="300" src="https://private-user-images.githubusercontent.com/45052050/472045514-6cca0be4-188a-4b54-b921-833cf26a39dd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM4MDEwNzMsIm5iZiI6MTc1MzgwMDc3MywicGF0aCI6Ii80NTA1MjA1MC80NzIwNDU1MTQtNmNjYTBiZTQtMTg4YS00YjU0LWI5MjEtODMzY2YyNmEzOWRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzI5VDE0NTI1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZlYzhhZTNmZDEyMWFjMzU5ZTUyOGVhMDJlZTg1NDk0MmI5NzAxYjYxMmIwOTc2Mjc1NjY1NmUyNWQ3YmJlNTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.c7GXrKZebgNI8tOqYtzJyVIRisKCfYcLsGSwlvd2ZpQ">
> 
> So, if we imagine that 3 nodes fit per page, we get node to page mapping such as: <img alt="image" width="300" height="300" src="https://private-user-images.githubusercontent.com/45052050/472046621-e58aa6c4-dc86-4bb8-a6be-7f5c17621bdf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM4MDEwNzMsIm5iZiI6MTc1MzgwMDc3MywicGF0aCI6Ii80NTA1MjA1MC80NzIwNDY2MjEtZTU4YWE2YzQtZGM4Ni00YmI4LWE2YmUtN2Y1YzE3NjIxYmRmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzI5VDE0NTI1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZiYTQwZjkwNmVjOGJhZDE1ZGU4ZDI4MGZmN2ExMTc4MDcwMTcwZTNjN2ZiOGU5NzFkMWY5NDA4YzFiOTRkNjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.uhNch6baZFtkyN6tVPhWWvQl9_F3tO6rxviKTXaFz6g">
> 
> But that's suboptimal, because many leaves now need 3 page jumps, and 2 would be enough. The subtree-accummulating logic tries to make it more like this: <img alt="image" width="300" height="300" src="https://private-user-images.githubusercontent.com/45052050/472047210-cb44832e-438b-4949-b255-1e147fe7f889.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM4MDEwNzMsIm5iZiI6MTc1MzgwMDc3MywicGF0aCI6Ii80NTA1MjA1MC80NzIwNDcyMTAtY2I0NDgzMmUtNDM4Yi00OTQ5LWIyNTUtMWUxNDdmZTdmODg5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzI5VDE0NTI1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ4ZGEzZjJiYzQwZDA3NmE1Y2U4NmE5MGQ5ODE5ZjBhZWI5ODFlNjU0MWY3M2QzMWFiM2UwOWZjNDA0MTU5NzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.HEJyiXy2YTzJtmQDZi9_h5G3wuRUPJmDHlY-OFijouw">
> 
> And for that, it wants to delay writing the left child of root (node 7 in first image) (which has a small live subtree, since its children were already flushed) until we complete the root and decide whether it can fit on one page with this child or not.

I see. It makes sense. But for a btree, this leveling is more natural.

michoecho replied:
> I see. It makes sense. But for a btree, this leveling is more natural.

Yes. It's natural for a btree, it needs a bunch of heuristics in a trie.

avikivity commented on sstables/trie/trie_writer.hh:
Limiting transition sizes and inlining them would improve performance (if you copy the whole thing unconditionally).

Clustering keys are limited to 64k and are typically much smaller. What about limiting transitions to 128 bytes?

michoecho replied:
We can keep this in mind as an optimization idea, but this is something that has to be tested under a benchmark. I'm not convinced it would improve performance. For example, in a partition index, the transition will almost always be 1 byte. If you inline a 128 byte buffer, you are wasting memory. This could be worse than a call to memcpy.

michoecho replied:
> Can you share your experience?

I just already played with this kind of optimizations (mostly by running `perf-sstable`) , and I did try eliminating some memcpy calls. And my conclusion was that, for the moment, this kind of optimization (i.e. making `memcpy` constant-sized) isn't worthwhile. 

The trie construction logic is heavy in general. It seems it's going to slow down things like flushes and easy compactions by about 5-10%.

(That's probably the main disadvantage of this format: the writer logic is fairly complicated and branchy, both for the programmer and for the CPU. In a simpler index, you could just append keys to a file and hoist one key every 4 kiB to a higher tree level).

avikivity replied:
Seems okay to me and we can defer optimizations until later.

michoecho commented on sstables/trie/trie_writer.hh:
Yes. The node's allocations are discarded after it's written out, but we still need to keep it's first transition byte (because its included in the parent), so we stuff it into some member variable that's not used anymore. It didn't have to be `_transition_meta`, but it seemed appropriate.

avikivity commented on sstables/trie/trie_writer.hh:
std::flat_set?

avikivity replied:
Is the comparator valid? I don't think it gives a total order. What happens if two separate items compare equivalent?

michoecho replied:
> Is the comparator valid? I don't think it gives a total order. What happens if two separate items compare equivalent?

Hmm, 
The comparator used for sorting, i.e.
```
        bool operator()(ptr<writer_node> a, ptr<writer_node> b) const {
            return std::make_pair(a->_branch_size + a->_node_size, a->_transition[0])
                 < std::make_pair(b->_branch_size + b->_node_size, b->_transition[0]);
        }
```
is total, because `_transition[0]` is different for each child, so it breaks ties.

The comparator used for lookup, i.e.
```
        bool operator()(int64_t a, ptr<writer_node> b) const {
            return a < (b->_branch_size + b->_node_size).value;
        }
```
means that `a` should be equivalent (in the sorting comparator) to a `ptr<writer_node>` with `b->_branch_size + b->_node_size` equal to `a` and `_transition[0]` equal to `-∞`. Right?

Is there anything which could distinguish between the `int64_t a` and the equivalent `(a, -∞)` sorting comparator pair?

avikivity commented on sstables/trie/trie_writer.hh:
Ok, please make it explicit.

michoecho commented on sstables/trie/trie_writer.cc:
```diff
@@ -0,0 +1,11 @@
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
```
Yeah, yeah. I'll push a new revision of the PR soon.

michoecho replied:
Fixed.

michoecho commented:
v2:
- Changed the namespace from `trie` to `sstables::trie`.
- Added `trie_writer.cc` to `CMakeLists.txt`
- Split `trie_writer.hh` into several files: `writer_node.hh` (`class bump_allocator` and `class writer_node`), `writer_node.impl.hh` (definition of the template members of `writer_node`), `common.hh` (`trie` logger, `const_bytes`, `developer_build`).
  (This splitting is guided by 
- Moved non-templated definitions to `trie_writer.cc`.
- Fixed the nonsense `#define TRIE_SANITIZE_BUMP_ALLOCATOR developer_build` line. (Now it's `#define TRIE_SANITIZE_BUMP_ALLOCATOR 0`, and there's a `static_assert` that makes `TRIE_SANITIZE_BUMP_ALLOCATOR` dependent on `developer_build`).
- Fixed the cosmetic problems noticed by the bots.
- Turned the double-role `writer_node::_transition_meta` into a union of `_transition_length` and `_first_transition_byte`.
- Added the requested comments.

michoecho commented on sstables/trie/trie_writer.hh:
Fixed.

michoecho commented:
@avikivity Re-review, please. Check if the changes are satisfactory and that I didn't skip some suggestion you insist on.

michoecho commented:
v2.01: I just realized that in the last push I left trace logs enabled in the new test, which slows it down by several seconds. So in the newest push I just lowered the log level back.

This little maneuver is going to cost us another 18 hours of rerunning the whole CI.

michoecho commented:
> 🔴 CI State: FAILURE

https://github.com/scylladb/scylla-pkg/issues/5337
